### DISCUSSION TITLE
Resolving Python Version Conflict with Poetry: Downgrade from Python 3.11 to 3.10 Required for pyproject.toml Compatibility

### DISCUSSION THREAD

**Discussion by codeananda** [python-poetry/poetry#8805] [Q&A] [answered]
I got my client to git clone a repo and asked him to `poetry install --no-root` as per the [poetry docs](https://python-poetry.org/docs/basic-usage/). 

However, he had installed python 3.11 from python.org and the `pyproject.toml` requires `python = ">=3.10,<3.11"`, so it errored out. How to fix this?

I had no issues downgrading from python 3.11 to 3.10 during my development. Why?

Possible explanations:
- I have anaconda also installed
- I'm using PyCharm

I picked poetry because I thought it would be an effortless way for my client to set up his environment but have fallen at the first hurdle.

What's the recommended way to deal with this?

**Comment by radoering** [comment]
Poetry does not manage Python installations so if your project requires Python 3.10, Python 3.10 has to be available on the target system. If there are multiple Python versions, you can tell Poetry which version to use with the [env use command](https://python-poetry.org/docs/managing-environments/#switching-between-environments).

Alternatively, you can of course also allow Python 3.11 by editing your `pyproject.toml`.

**Reply by codeananda** [reply]
Oh right. Is there a plan to support this? Coming from conda, it feels like a massive step back that it can't. I've been modifying the supported python versions in pyproject.toml like any other package. Probably because I had them installed via conda?

**Reply by radoering** [reply, marked as answer]
> Oh right. Is there a plan to support this?

There are no plans in the foreseeable future. There is already pyenv, which can be used together with Poetry.

**Reply by codeananda** [reply]
For anyone else looking for a single package manager to handle everything (including python installation), check out [pixi](https://github.com/prefix-dev/pixi). It's basically poetry with poetry's 2 major problems (1. it doesn't install python, 2. it can't install non-python packages) fixed. 

Or [uv](https://github.com/astral-sh/uv) which is poetry but faster and installs python and is becoming the most popular package manager for pure python projects

**Reply by Ivan-SB** [reply]
I do understand the project goals but I've a poetry project where I should downgrade the python requirement. My system has only python 3.13 installed globally and poetry installed from apt, what I think I could do would be install a lower version globally, edit pyproject.toml, run poetry update from inside the project dir but outside the project virtualenv. I'm just guessing, but could you suggest a better way, possibly avoiding installing another python version globally? thanks

**Reply by codeananda** [reply]
@Ivan-SB this is a HUGE downside of poetry. I am baffled why the maintainers do not want to give poetry the ability to handle python installs 🤷‍♂️

Options

1. Install the required python version manually (i.e. from python website)
2. Install required python version with pyenv
3. Switch to uv for package management

You need to install a version of python that the project supports. There is no way around that.

**Reply by radoering** [reply]
Nowadays, you can also try https://python-poetry.org/docs/cli/#python

**Reply by codeananda** [reply]
@radoering  That link took me here for some reason https://python-poetry.org/docs/cli/#script-project

But I see from the [2.1 release notes](https://python-poetry.org/blog/announcing-poetry-2.1.0/#experimental-commands-to-manage-python-installations) that there are now experimental commands to manage python installations!! Wow this is great news

<img width="896" alt="image" src="https://github.com/user-attachments/assets/7da0ee57-2df7-4a99-acda-248163cae2be" />
